File: performance-comparison/benchmarks/rust-workspace/src/bin/alloc_churn.rs
// alloc_churn -- allocate and free one small heap object per iteration
// (Box, matching the malloc/free round-trip), summing the stored value.
use std::env;

fn churn(n: i64) -> i64 {
    let mut sum: i64 = 0;
    for i in 0..n {
        let p = Box::new(i);
        sum += *p;
        drop(p);
    }
    sum
}

fn main() {
    let n: i64 = env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(10000);
    println!("{}", churn(n));
}

File: performance-comparison/benchmarks/haskell-project/src/AllocChurn.hs
-- alloc_churn -- one small heap object per iteration (an IORef box stands
-- in for the malloc/free round-trip; GC reclaims it, which IS the
-- allocator story being measured in a GC'd column).
{-# LANGUAGE BangPatterns #-}
import System.Environment (getArgs)
import Data.IORef
main :: IO ()
main = do
  args <- getArgs
  let n = case args of (a:_) -> read a; _ -> 10000 :: Int
      go !i !total
        | i >= n = return total
        | otherwise = do
            ref <- newIORef i
            v <- readIORef ref
            go (i + 1) (total + v)
  total <- go 0 (0 :: Int)
  print total
